### PR TITLE
Add vqa

### PR DESCRIPTION
**Pull request by henryj18**
This PR includes changes on image-caption and vqa code based off add_text_modality

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -16,4 +16,18 @@ See [Issues](https://github.com/ManifoldRG/NEKO/issues) for more information. If
...
+- Define a new task type in `gato/tasks/task.py`
+- Define your new task under the `gato/tasks` folder (refer to the tasks already implemented here, you should have a good idea about what you need to do):
+    - Open and load the datasets for the task, process the data into the format to be accepted by model training
+    - Define the function to sample data from this task for training
+    - Define a function to evaluate the performance of this task
+- Define the function for inference/predicton specific to this task in `gato/policy/gato_policy.py`. The rest of this file defines the training model for NEKO, and most likely you do not need to touch that part
+- Add a few lines of code in `gato/training/trainer.py` to sample data for the new task and add to each batch for training
+- Add a few lines of code in `train.py` to input the command-line args specific to your task and to instantiate an instance of your task
+
```
Comment (bhavul) [inline CONTRIBUTING.md:30]: We will need another entry -- we should probably add an independent evaluation code in `eval.py`

File: gato/policy/gato_policy.py [hunk 2]
```diff
@@ -340,7 +347,9 @@ def tokenize_input_dicts(self, inputs: list):
             batch_tokens = torch.cat(
                 [
                     tokens for tokens in
-                    [text_tokens, image_tokens, continuous_tokens, discrete_tokens, separator_tokens, continuous_action_tokens, discrete_action_tokens]
+                    # Swap the order of image_tokens and text_tokens (GATO paper places test tokens before image tokens) for the image-caption task,
+                    # caption (which a text_tokens) should follow image instead of preceding image, so that image data can auto-regressively predict caption tokens
```
Comment (bhavul) [inline gato/policy/gato_policy.py:351]: I'm unsure if this is actually how it works. 
@daniellawson9999 care to comment?

File: gato/policy/gato_policy.py [hunk 3]
```diff
@@ -458,6 +467,88 @@ def predict_text(self, batch_dict, max_length=20, deterministic=True):
...
+
+        response_tokens = []
+
+        for idx in range(max_length): # idx can be viewed as the index of the 'next_token' within the list of such generated response tokens
+
+            # Include 'image_embeddings' instead of 'images' below to avoid re-calculating the same image embedding in every loop
+            # Note: the following 'text' is actually text token IDs instead of actual text. 
+            # We were using actual text before (and at that time tokenize_input_dicts(...) function also took actual text), and found an error 
+            # due to an edge case. The edge case is that the generated tokens could contain a sequence of .... (periods) or ,,,, (commas), 
+            # and if actual text is used in the following 'text', the tokenizer will treat the sequence as one token (which is not the desired behavior)
```
Comment (bhavul) [inline gato/policy/gato_policy.py]: The comments related to `text` can be removed. It wouldn't be helpful to someone in future.

File: gato/policy/gato_policy.py [hunk 4]
```diff
@@ -458,6 +467,88 @@ def predict_text(self, batch_dict, max_length=20, deterministic=True):
...
+            # of the corresponding response. When predicting the response from an image, we pass the image embeddings of the image tokens through 
+            # the transformer, and then auto-regressively predict the next tokens, the context info of the image is conveyed through the 
+            # self-attention mechanism. The first text token of the generated response is the "next token" predicted by the last image token 
+            # in the sequence, and the second text token is predicted by the first text token, so on and so forth. 
+            # In the following line of code:
+            #   n_patches - 1 + len(prompt_tokens) + idx is the index of the token in the sequence that will predict the next text token of the generated response
+            #   squeeze the logits from shape (1, 1, vocab_size) to shape (vocab_size) and restrict the choice to the range 
+            #       within text tokens. The complete vocab contains non-text toksns also, we need to to exclude them from the choice
+            next_token_logits = logits[:, n_patches -1 + len(prompt_tokens) + idx, :].squeeze()[:self.text_tokens]     # shape (self.text_tokens)
```
Comment (bhavul) [inline gato/policy/gato_policy.py]: This is probably incorrect slicing. I will debug to check but it looks quite different from what's happening within text.

File: gato/policy/gato_policy.py [hunk 5]
```diff
@@ -458,6 +467,88 @@ def predict_text(self, batch_dict, max_length=20, deterministic=True):
...
+            if deterministic:
+                next_token = torch.argmax(next_token_logits).item()
```
Comment (bhavul) [inline gato/policy/gato_policy.py:520]: This may require `dim` as well

File: gato/tasks/caption_task.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+class CaptionTask(Task): 
+    def __init__(self, task_type: TaskTypeEnum, tokenizer_model:str, caption_dataset, train_data, test_data = [],
+                 test_data_prop = 0.1, test_data_mask_file = None):
+        """
+        task_type should be CAPTION
+        caption_dataset is the directory for all of the data (training and test)
+        train_data and test_data are list of sub_diretories under caption_dataset, with each directory containing 
+        one training or test dataset which is composed of multiple .tar files downloaded with img2dataset. 
+        Each tar file contains multiple bundles, with each bundle containing one .jpg, one txt and one json file.
```
Comment (bhavul) [inline gato/tasks/caption_task.py:28]: We should to have a better standard here, so as we scale up machines, it's easy to try a different dataset.

File: gato/tasks/caption_task.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            'loss': avg_loss,
+            'perplexity': perplexity.item()
+        }
+        return metrics
+
+# test code
+if __name__ == '__main__':
+    # replace the following directory with your data directory
+    task = CaptionTask(task_type = 'caption', tokenizer_model = 'gpt2', caption_dataset = '/home/<user name>/Git/NEKO/Caption_data', 
+                       train_data = ['train'], test_data = ['test'], test_data_prop = 0.1)
```
Comment (bhavul) [inline gato/tasks/caption_task.py:165]: The path should come from CLI args instead of hardcoding here

**Review by bhavul — COMMENTED** [review]
Have checked caption_task, will have a look at vqa soon as well.
Need to try out code with debugger to make sure slicing is correct.

File: CONTRIBUTING.md (hunk 1 above)
Comment (henryj18) [reply CONTRIBUTING.md:30]: Sure, let's discuss about it, just came back from vacations, sorry for the delay

**Comment by henryj18** [comment]
Merge add_vqa to master
